### PR TITLE
CIF-287 - PROPOSAL: Debugging mode

### PR DESCRIPTION
**Pull request by herzog31**
**NOTE:** This is just a proposal, open for discussion. Please do not merge.

The motivation for these changes are to make developers' lives easier and to enable deeper insights for testing.

### 1. Enable Debug mode
You can enable the debug mode by setting the `DEBUG` parameter either as query or form parameter.

### 2. OpenWhisk Activation ID
If the debug mode is enabled, the action will return it's activation Id as response header in the `OW-Activation-Id` field. Exposing an activation ID to the outside is no security concern, since it's only useful if you have actually access to the OpenWhisk namespace.

#### Example
Request: `https://runtime.adobe.io/api/v1/web/namespace/magento/getCategories?DEBUG=1`
Response Header: `OW-Activation-Id: c58858d6a7d642f78858d6a7d682f71d`

### 3. Backend Requests
Every request made to the commerce engine (Magento in this case) is logged and visible in the activation log. In comparison to the previous approach with the performance headers, the information is not exposed to the public and only available with access to the OpenWhisk namespace.

I also looked into alternative approaches. In comparison to the headers, the logging is much simpler and only requires changes at one location (where the request is made). All other metrics like action invocations, sequence duration etc. can be retrieved from the activation logs. Another approach is the usage of OpenWhisk triggers which can be bound to actions. This approach however has the disadvantage that the `openwhisk` node package communicates with the OpenWhisk API via HTTP calls. This comes with potential performance drawbacks. Also to actually process the performance data, for each trigger an action has to be called which results in a high load on the OpenWhisk instance.

#### Example
Request: `https://runtime.adobe.io/api/v1/web/namespace/magento/postCartEntry?DEBUG=1&productVariantId=eqbisumas-10&quantity=2`
Activation log: 
```
"logs": [
    "2018-08-10T10:43:44.261868285Z stdout: BACKEND-CALL POST http://hostname/rest/V1/guest-carts 556188.263 PASS",
    "2018-08-10T10:43:45.140676575Z stdout: BACKEND-CALL POST http://hostname/rest/V1/guest-carts/481ce6af452e567f3212b5fed7fa2337/items 878420.555 PASS",
    "2018-08-10T10:43:45.97952998Z  stdout: BACKEND-CALL GET http://hostname/rest/V1/guest-aggregated-carts/481ce6af452e567f3212b5fed7fa2337?productAttributesSearchCriteria[filter_groups][0][filters][0][field]=attribute_code&productAttributesSearchCriteria[filter_groups][0][filters][0][value]=color&productAttributesSearchCriteria[filter_groups][0][filters][1][field]=attribute_code&productAttributesSearchCriteria[filter_groups][0][filters][1][value]=size 838461.798 PASS"
]
```

### CODE REVIEW

File: src/common/MagentoClientBase.js [hunk 1]
```diff
@@ -113,13 +117,49 @@ class MagentoClientBase {
...
-    
+
+    /**
+     * Wraps a request and prints out debug information to the log.
+     * 
+     * @param {Object} options  Request options.
+     */
+    _profileRequest(options) {
+        let s = process.hrtime();
+        let logRequest = this._logRequest.bind(this);
```
Comment (masodu) [inline src/common/MagentoClientBase.js]: why is this here for?

Comment (herzog31) [reply src/common/MagentoClientBase.js]: Inside the promise.then, `this` no longer refers to the `MagentoClientBase` class, but the promise context. So you either store the class context inside a different variable (e.g. `that`) or you can store a copy of the function that is already bound to the class context in a different scope and use it from the promise.

Comment (masodu) [reply src/common/MagentoClientBase.js]: and where is the class context used in the `_logRequest`? Is the `process.hrTime()` needing it ?

Comment (herzog31) [reply src/common/MagentoClientBase.js]: No, it would work without the bind as well. If someone would change the logRequest method however and add a `this`, I think he would expect it to work, since the scoping might not be obvious.

Comment (masodu) [reply src/common/MagentoClientBase.js]: OK, then please add a simple inline comment to the bind statement.

**Comment by mhaack** [comment]
Sorry to jump in here lately. I like to see `decorateActionForSequence` going away.

But the current approach might be a security problem? And as written has performance impact as we. Can't we make this a default OW config param which can not be overwritten via URL. I think this should be some config the deployer of the service decides, not the one who calls the service.

**Comment by herzog31** [comment]
@mhaack where did you read about the performance impact? This approach just adds some console.log statements which don't have any performance impact.

Regarding the security concern: That's how OpenWhisk works. For OpenWhisk it doesn't matter whether you add a parameter as query, post, or config parameter. They all end up being in the `args` object. What we can do, is to filter query and post parameters as part of the REST API configuration. But this wouldn't change the current implementation. However, as I wrote, there is no data exposed other than the activation id, in debug mode.

**Comment by mhaack** [comment]
@herzog31 I think there might be some performance aspect since OW activation log will be bigger, hence more data must be processed OW internally.

Regarding the parameters & default parameters, I'm referring to https://github.com/apache/incubator-openwhisk/blob/master/docs/webactions.md#protected-parameters

**Comment by herzog31** [comment]
@mhaack Ok, makes sense to add this to protected parameters. I think I can add this to the serverless configuration to disable debugging by default and it has to be explicitly enabled in the serverless configuration to allow it.

What alternative do you propose? Where can we store that data?

**Comment by masodu** [comment]
@mhaack I think that the performance impact, if any, is minimal and not relevant for the scope of this feature. I would be more concerned about action responses like getCart which can be big. If this is really a concern should we ask runtime team? Or do you have some insights already ?

**Comment by herzog31** [comment]
@mhaack @masodu I think you’re right, Sorin. The responses of our actions are also stored in the activation logs. And the responses sometimes are very large. 1-3 lines of logging should be insignificant in comparison.
